### ISSUE TITLE
[New build - DEFAULT] 23w41a: Oxygen Life Support, Air-tightness, Force fields

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#5302]
This build introduce a capital element of survival: oxygen.

**Player changes:**
- An oxygen meter bar has been added to the UI
- the player will die when this empty out
- Automatically refill inside breathable atmosphere or inside a ship pressurized room.
- Die much faster when helmet is off in space
- Of course those can be toggled On / Off in creative mode

**Life Support blocks**
The life Support system will pressurize the room it is placed inside.
There is a nice screen displaying the status and even a rotating fan with some cool VFX!
The algoritm works like in real life: a few oxygen molecules are emitted and will bounce around to find the volume pressurized.
If a molecule ends up outside the ship bounds a leak is declared and pressurization is not achieved.
The downside is that the volume is a box, ie a L shaped room will potentially give oxygen to a volume outside.
You can check how it is working by selecting a Life Support brick with the link tool (molecules travel more slowly in preview mode).
It worked well on the ship I tested, and it will react automatically to any new blocks or moving door / mechanism triggered.
The tank reserve allows for a long uptime (2 hours) before a replenish is needed. 
Tank replenish either manually (not recommended for large ships), or automatically inside an atmosphere or near a station (it is magic).
As it would be too tedious to have to pipe them, they cannot be connected to pipes at all.
Again an option has been added inside the Entity screen to provide Oxygen ship-wide for creative or older ships.

**Force-Field**
The force field will project a wall preventing oxygen from leaking, but allowing player or ship to cross it.
Perfect for hangars!
The brick will raycast at the two extremity to determinate the height of the field.
It comes in standard and wedge variant.

As a bonus two creative new features have been added:
- The Air Vent brick can now play the gas VFX (like in the old days) if active
- A new Fan brick has been added: you can configure the fan speed. This is purely decorative.

Other changes:
- Can see atmospheric and oxygen density inside the holomap

Still to do:
- Depressurization event

**Comment by ZachZent** [comment]
Been curious how this would go down and I love the idea. A question. How many of the vent blocks do you need per ship/volume? I can't imagine on a 1km long cargo ship you would only need one, right?



> The tank reserve allows for a long uptime (2 hours) before a replenish is needed. Tank replenish either manually (not recommended for large ships), or automatically inside an atmosphere or near a station (it is magic).

I like the idea of absorbing o2 from the atmosphere and not needing to pipe up. However one part that may have been overlooked, stations. They wouldn't be going into atmo so the vent would need to be refilled manually. And if the station is very large... A suggestion, a "control unit" block. Something that can be linked to the vents for refills from one source. Perhaps that can be piped into for automation.

**Comment by Asnro** [comment]
I love the "bouncing molecule" interpretation of airtightness, it should allow child entities to be airtight, right? This would be great news for players who build with lots of child entities!

However, a concern: if the life support block can't be piped into, what's the point of oxygen tanks and pipes? Can/could the life support blocks draw oxygen from tanks on the same entity, even if piping isn't allowed?

**Comment by tsunamayo** [comment]
Yeah maybe indeed for player station a magic connection to an oxygen tank would be needed. I will see what I can do. Yes this works great with children entity, the only downside is the box (as I am not longer using voxels...)

@ZachZent so one life support can "support" a 25000m3 volume, so the large hangar from SURGE on the tutorial needs two.
Cheers

**Comment by AlienXtream** [comment]
i was wondering. these particles. how many of them are there and could you, say, detect when a player hits/overlaps a particle relatively cheeply? if so, perhaps you could actually have a little timer on the player that gets reset when the player hits one of these particles. a "breath timer" if you like. actually, you wouldn't even need that. just have the players approximate head position do a simple check for particles in a small area around it every second or so. if none are detected for 30s the player is either not in an atmo or there isn't enough air in there anyway. i am not sure what you mean about piping? do you mean to refill the players air or for the vent to work? im a little unclear on that. i assume you mean that logistics pipes can not be connected but that the link tool can connect a vent to a tank?

**Comment by ProPeach** [comment]
Oh wow, what an awesome surprise! Very impressive that it works with child entities, most other games seem to struggle with that (I suppose because they are using voxels?) The visual of the particles is very cool too.

Huge thanks for the return of the gas VFX panel!! I've missed it a lot lol

I haven't had a chance to test it yet, but do you think the airtightness will work with planet/asteroid terrain? Say if you built an underground base on a moon or asteroid and had one wall as exposed rock, would you be able to pressurise that room? If not we can always cheat a little and hide some hull in the ground I suppose lol

**Comment by Briaireous** [comment]
@tsunamayo it appears that the prefab glass canopies leak regardless of how it is built.

the sloped one works fine. 
![Starship EVO 10_13_2023 11_22_52 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/aee57203-a536-4e78-891d-b47681028935)

the flat one, not so much even if it is on the main body
![Starship EVO 10_13_2023 11_20_58 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/7f4303d0-555c-40bb-81c7-c70f7b90242e)
![Starship EVO 10_13_2023 11_20_47 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/de073fda-fa45-4f0c-9acf-13b4da24dccd)
![Starship EVO 10_13_2023 11_20_52 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/4db7f551-22d4-4150-8f6b-e82fdd87851b)

**Comment by Briaireous** [comment]
![Starship EVO 10_13_2023 11_02_17 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/a34c0a40-da62-41f0-b02b-574b5c39eb57)
![Starship EVO 10_13_2023 11_15_39 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/244d6aa4-3a23-4d79-8130-c9733349a220)
![Starship EVO 10_13_2023 11_15_22 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/6029f4ab-d767-4dc6-b29c-30216923ac36)
![Starship EVO 10_13_2023 11_15_13 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/555205bb-25b0-449a-9da1-afc585be6bb2)
![Starship EVO 10_13_2023 11_14_59 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/4b70e9d7-756e-4a4c-baf4-377d8dbca9b3)
![Starship EVO 10_13_2023 11_14_37 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/d80eaaed-d9e9-413a-a682-b0d9a8479ad9)
![Starship EVO 10_13_2023 11_14_23 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/419eceed-18e5-49ec-876b-6bcead9c5ddb)
![Starship EVO 10_13_2023 11_14_11 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/d8848545-aa04-4001-85bc-bd03532ad697)
![Starship EVO 10_13_2023 11_13_38 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/611b8f86-747c-4590-a2cb-3d5e0cfbbbd4)
![Starship EVO 10_13_2023 11_13_24 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/2f420755-ec1e-4493-9aba-20bd0f46ecf3)
![Starship EVO 10_13_2023 11_12_54 AM](https://github.com/tsunamayo/Starship-EVO/assets/23586866/d983c385-3cf5-4955-bb8b-b2fe8a35a706)

**Comment by ProPeach** [comment]
> also quick question about the vent's sound effect, do you think we could have the option to change it to be more of a hissing sound like in classic star wars when whatever ship lands and steam shoots out with a hiss?


100% support this! :D
A good ***pshhhhhhhhh*** of escaping air would be perfect

**Comment by ultrak2k** [comment]
Hmm, I'm unsure on air refilling being just magic instead of having a pipe system, especially since there might be situations where you have a large ship that needs  an air refill but has to get it from not either a station or a planet that has oxygen in it's atmosphere, at least having the option to pipe up oxygen would be nice. I'm unsure how oxygen would work on stations as in, how would they get refilled? Would ships need to resupply them from atmospheres automatically? That would be quite interesting, though surely docking ports and pipes would be very fitting for that type of thing too.

**Comment by Pawnp** [comment]
Is there any progress with the suggestion on disabling the ship rename and 30 minute reminder? https://github.com/tsunamayo/Starship-EVO/issues/5197

**Comment by FreeLikeGNU** [comment]
Air tightness seems to be random when using curved glass and arches, I have to recheck air seal with link tool multiple times without changing the structure until it it reads "nominal" 

<img width="1916" height="1161" alt="Image" src="https://github.com/user-attachments/assets/f241078b-387f-4e6a-b34c-17eac5eae1b6" />
